sdk/storage/azure-storage-blobs/test/ut/blob_sas_test.cpp

```diff
@@ -317,301 +217,535 @@ namespace Azure { namespace Storage { namespace Test {
+
+    auto keyCredential
+        = _internal::ParseConnectionString(StandardStorageConnectionString()).KeyCredential;
+
+    auto blobClient = *m_blockBlobClient;
+    const std::string blobName = m_blobName;
+
+    Sas::BlobSasBuilder blobSasBuilder;
+    blobSasBuilder.Protocol = Sas::SasProtocol::HttpsAndHttp;
+    blobSasBuilder.StartsOn = sasStartsOn;
+    blobSasBuilder.ExpiresOn = sasExpiresOn;
+    blobSasBuilder.BlobContainerName = m_containerName;
+    blobSasBuilder.BlobName = blobName;
+    blobSasBuilder.Resource = Sas::BlobSasResource::Blob;
+    blobSasBuilder.SetPermissions(Sas::BlobSasPermissions::All);
+    blobSasBuilder.IPRange = "0.0.0.0-0.0.0.1";
+
+    auto sasToken = blobSasBuilder.GenerateSasToken(*keyCredential);
+    VerifyBlobSasNonRead(blobClient, sasToken);
+
+    blobSasBuilder.IPRange = "0.0.0.0-255.255.255.255";
+    sasToken = blobSasBuilder.GenerateSasToken(*keyCredential);
+    VerifyBlobSasRead(blobClient, sasToken);
+  }
+
+  TEST_F(BlobSasTest, AccountSasService)
+  {
+    auto sasStartsOn = std::chrono::system_clock::now() - std::chrono::minutes(5);
+    auto sasExpiresOn = std::chrono::system_clock::now() + std::chrono::minutes(60);
+
+    auto keyCredential
+        = _internal::ParseConnectionString(StandardStorageConnectionString()).KeyCredential;
+
+    auto blobClient = *m_blockBlobClient;
+
+    Sas::AccountSasBuilder accountSasBuilder;
+    accountSasBuilder.Protocol = Sas::SasProtocol::HttpsAndHttp;
+    accountSasBuilder.ExpiresOn = sasExpiresOn;
+    accountSasBuilder.Services = Sas::AccountSasServices::Blobs;
+    accountSasBuilder.ResourceTypes = Sas::AccountSasResource::All;
+    accountSasBuilder.SetPermissions(Sas::AccountSasPermissions::All);
+    accountSasBuilder.Services = Sas::AccountSasServices::Files;
+
+    auto sasToken = accountSasBuilder.GenerateSasToken(*keyCredential);
+    VerifyBlobSasNonRead(blobClient, sasToken);
+
+    accountSasBuilder.Services = Sas::AccountSasServices::All;
+    sasToken = accountSasBuilder.GenerateSasToken(*keyCredential);
+    VerifyBlobSasRead(blobClient, sasToken);
+  }
+
+  TEST_F(BlobSasTest, AccountSasResourceTypes)
+  {
+    auto sasStartsOn = std::chrono::system_clock::now() - std::chrono::minutes(5);
+    auto sasExpiresOn = std::chrono::system_clock::now() + std::chrono::minutes(60);
+
+    auto keyCredential
+        = _internal::ParseConnectionString(StandardStorageConnectionString()).KeyCredential;
+
+    auto blobClient = *m_blockBlobClient;
+
+    Sas::AccountSasBuilder accountSasBuilder;
+    accountSasBuilder.Protocol = Sas::SasProtocol::HttpsAndHttp;
+    accountSasBuilder.ExpiresOn = sasExpiresOn;
+    accountSasBuilder.Services = Sas::AccountSasServices::Blobs;
+    accountSasBuilder.ResourceTypes = Sas::AccountSasResource::All;
+    accountSasBuilder.SetPermissions(Sas::AccountSasPermissions::All);
+    accountSasBuilder.ResourceTypes = Sas::AccountSasResource::Service;
```

Comment:
Why is `accountSasBuilder.ResourceTypes` assigned twice here?

Reply:
fixed.